### PR TITLE
[WIP] add container informer

### PR DESCRIPTION
**Pull request by halegreen**
### V. Checklist

- [ ] I have written necessary docs and comments
- [ ] I have added necessary unit tests and integration tests
- [ ] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/statesinformer/container_informer.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+}
+
+func NewContainerInformer() *containerInformer {
+	p, err := pleg.NewPLEG(system.Conf.CgroupRootDir)
+	if err != nil {
+		klog.Fatalf("failed to create PLEG, %v", err)
+	}
+
+	informer := &containerInformer{
+		containerMap:       make(map[string]map[string]*ContainerMeta),
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/container_informer.go:65]: divide into two maps may be easy understanding.
map[podUID][]*ContainersMeta
map[ContainerHashID]*ContainerMeta

File: pkg/koordlet/statesinformer/container_informer.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	ret := []*ContainerMeta{}
+	for _, containers := range c.containerMap {
+		for _, cMeta := range containers {
+			ret = append(ret, cMeta)
+		}
+	}
+	return ret
+}
+
+func (c *containerInformer) GeSandboxContainer(podUID string) *ContainerMeta {
```
Comment (zwzhang0107) [inline pkg/koordlet/statesinformer/container_informer.go:206]: we need to separate whether a container is pod sandbox type.

File: pkg/koordlet/statesinformer/api.go [hunk 3]
```diff
@@ -16,16 +16,70 @@ limitations under the License.
...
+
+type InformerGetter interface {
+	GetInformer(name informerName) Informer
+}
+
+type Informer interface {
+	Run(stopCh <-chan struct{}) error
+	HasSynced() bool
+
+	GetNode() *corev1.Node
```
Comment (saintube) [inline pkg/koordlet/statesinformer/api.go:39]: perhaps the generic Informer does not require the following methods `GetNode`, `GetNodeSLO`, ... `GetContainerByPodUID`

File: pkg/koordlet/statesinformer/api.go [hunk 4]
```diff
@@ -16,16 +16,70 @@ limitations under the License.
...
+	GetAllPods() []*PodMeta
+	GetNodeTopo() *topov1alpha1.NodeResourceTopology
+	RegisterCallbacks(objType RegisterType, name, description string, callbackFn UpdateCbFn)
+	GetContainer(containerHashID string) *ContainerMeta
+	GetAllSandboxContainers() []*ContainerMeta
+	GeSandboxContainer(podUID string) *ContainerMeta
+	GetContainersByPodUID(podUID string) []*ContainerMeta
+}
+
+type informerName string
```
Comment (saintube) [inline pkg/koordlet/statesinformer/api.go:50]: this type should be public

File: pkg/koordlet/statesinformer/states_informer.go [hunk 5]
```diff
@@ -226,6 +225,22 @@ func (s *statesInformer) GetAllPods() []*PodMeta {
 	return podsInformer.GetAllPods()
 }
 
+func (s *statesInformer) GetContainer(containerHashID string) *ContainerMeta {
```
Comment (saintube) [inline pkg/koordlet/statesinformer/states_informer.go:228]: No need to declare these methods for the statesInformer.
And other modules may use the statesInformer like:

```go
si := informerGetter.GetInformer(statesinformer.StatesInformerName).(StateInformer)
```

File: pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector.go [hunk 6]
```diff
@@ -153,7 +153,7 @@ func (b *beResourceCollector) getBECPURealMilliLimit() (int, error) {
 
 func (b *beResourceCollector) getBECPURequestSum() resource.Quantity {
 	requestSum := int64(0)
-	for _, podMeta := range b.statesInformer.GetAllPods() {
+	for _, podMeta := range b.statesInformer.(*statesinformer.StatesInformerImpl).GetAllPods() {
```
Comment (zwzhang0107) [inline pkg/koordlet/metricsadvisor/collectors/beresource/be_resource_collector.go:156]: we can use the `PodsInformerPlugin` here, and beResourceCollector can be inited by statesInformer.GetInfromerPlugin()

**Comment by zwzhang0107** [comment]
hi @ halegreen, is there any new idea about this?

**Pull request closed without merging**